### PR TITLE
Add support for build and push to ECR from local Dockerfile

### PR DESCRIPTION
**Pull request by rli**
image

https://user-images.githubusercontent.com/742829/114238996-50f02380-993a-11eb-8515-a2343df1a87d.mp4

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushTagToRepositoryAction.kt [hunk 1]
```diff
@@ -150,20 +209,155 @@ internal class PushToEcrDialog(
...
+        }
+    }
+
+    private fun buildDockerfileActions(runConfigModel: CollectionComboBoxModel<DockerRunConfiguration>, textComponent: ExtendableTextField) {
+        val editExtension = ExtendableTextComponent.Extension.create(
+            AllIcons.General.Inline_edit, AllIcons.General.Inline_edit_hovered,
+            message("ecr.dockerfile.configuration.edit")
+        ) {
+            runConfiguration?.let {
+                RunManager.getInstance(project).findSettings(it)?.let { settings ->
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushTagToRepositoryAction.kt:260]: We can set a max size so the panel isn't so wide:
```maximumSize = Dimension(400, 32767)```

File: jetbrains-core/it-202-203/software/aws/toolkits/jetbrains/services/ecr/EcrPushIntegrationTest.kt [hunk 2]
```diff
@@ -51,21 +59,56 @@ class EcrPushIntegrationTest {
...
+            pushImage(projectRule.project, ecrLogin, pushRequest)
+        }
+
+        assertThat(
+            ecrClient.describeImages {
+                it.repositoryName(remoteRepo.repositoryName)
+            }.imageDetails()
+        ).anySatisfy {
+            assertThat(it.imageTags()).contains(remoteTag)
+            // would check the digest matches what we've uploaded, but it appears to change upon upload to ECR
```
Comment (hunterwerlla) [inline jetbrains-core/it-202-203/software/aws/toolkits/jetbrains/services/ecr/EcrPushIntegrationTest.kt:86]: nit: update the comment to mention it's because the image is encrypted/link the doc?

File: jetbrains-core/it-202-203/software/aws/toolkits/jetbrains/services/ecr/EcrPushIntegrationTest.kt [hunk 3]
```diff
@@ -51,21 +59,56 @@ class EcrPushIntegrationTest {
...
+
+    @Test
+    fun testPushFromDockerfile() {
+        val remoteTag = UUID.randomUUID().toString()
+
+        val dockerfile = folder.newFile()
+        dockerfile.writeText(
+            """
+                # arbitrary base image with a shell
+                FROM public.ecr.aws/lambda/provided:latest
```
Comment (hunterwerlla) [inline jetbrains-core/it-202-203/software/aws/toolkits/jetbrains/services/ecr/EcrPushIntegrationTest.kt:98]: nit: we could do `FROM scratch` and not even pull an image

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushTagToRepositoryAction.kt [hunk 4]
```diff
@@ -40,12 +63,15 @@ import software.aws.toolkits.jetbrains.utils.notifyError
 import software.aws.toolkits.jetbrains.utils.ui.blankAsNull
 import software.aws.toolkits.jetbrains.utils.ui.selected
 import software.aws.toolkits.resources.message
+import javax.swing.JComponent
+import javax.swing.JTextField
+import javax.swing.plaf.basic.BasicComboBoxEditor
 
 class PushTagToRepositoryAction :
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushTagToRepositoryAction.kt:70]: nit: revert the name back to what it was before since it does both

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushTagToRepositoryAction.kt [hunk 5]
```diff
@@ -131,15 +160,45 @@ internal class PushToEcrDialog(
...
 
     override fun createCenterPanel() = panel {
-        // attach a group so combo boxes grow/shrink together
-        val selectorGroup = "selectors"
+        lateinit var fromLocalImageButton: JBRadioButton
+        lateinit var fromDockerfileButton: JBRadioButton
+
+        buttonGroup(::type) {
+            row {
+                fromLocalImageButton = this@row.radioButton("Local Image", BuildType.LocalImage).component
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushTagToRepositoryAction.kt:168]: nit: need to do localization still

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushTagToRepositoryAction.kt [hunk 6]
```diff
@@ -131,15 +160,45 @@ internal class PushToEcrDialog(
...
+                fromDockerfileButton = this@row.radioButton("Dockerfile", BuildType.Dockerfile).component
+            }
+        }
+
+        val imageSelectorPanel = localImageSelectorPanel()
+        val dockerfilePanel = dockerfileConfigurationSelectorPanel()
+
+        row {
+            // TODO: panel is still jumping around
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushTagToRepositoryAction.kt:177]: What does this mean?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushTagToRepositoryAction.kt [hunk 7]
```diff
@@ -150,20 +209,155 @@ internal class PushToEcrDialog(
...
+                .constraints(grow)
                 .withErrorOnApplyIf(message("ecr.image.not_selected")) { it.selected() == null }
         }
+    }
 
-        row(message("ecr.push.remoteTag")) {
-            textField(::remoteTag)
-                .sizeGroup(selectorGroup)
-                .withErrorOnApplyIf(message("ecr.tag.not_provided")) { it.blankAsNull() == null }
+    private fun dockerfileConfigurationSelectorPanel() = panel {
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushTagToRepositoryAction.kt:217]: curious why this is a function if we call it in init anyway?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushTagToRepositoryAction.kt [hunk 8]
```diff
@@ -150,20 +209,155 @@ internal class PushToEcrDialog(
...
+
+        else -> throw IllegalStateException()
+    }
+
+    private enum class BuildType {
+        LocalImage, Dockerfile
+    }
+}
+
+// TODO: unify utils with other PR
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushTagToRepositoryAction.kt:340]: to fulfill this TODO, just dump it in the same package and I'll fix it on the other side when this merges

File: jetbrains-core/it-202-203/software/aws/toolkits/jetbrains/services/ecr/EcrPushIntegrationTest.kt (hunk 3 above)
Comment (rli) [reply jetbrains-core/it-202-203/software/aws/toolkits/jetbrains/services/ecr/EcrPushIntegrationTest.kt:98]: we could, but this is slightly cleaner (unless you know of a way to guarantee that a new image will be always be built)

File: jetbrains-core/it-202-203/software/aws/toolkits/jetbrains/services/ecr/EcrPushIntegrationTest.kt (hunk 2 above)
Comment (rli) [reply jetbrains-core/it-202-203/software/aws/toolkits/jetbrains/services/ecr/EcrPushIntegrationTest.kt:86]: updated test to inspect the manifest for the image id

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushTagToRepositoryAction.kt (hunk 7 above)
Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushTagToRepositoryAction.kt:217]: no particular reason, it was just getting too hard to follow as a single block

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushTagToRepositoryAction.kt (hunk 1 above)
Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushTagToRepositoryAction.kt:260]: opted for `growPolicy(GrowPolicy.MEDIUM_TEXT)` instead

Comment (hunterwerlla) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushTagToRepositoryAction.kt]: That works better, cool

**Comment by hunterwerlla** [comment]
Tried it out locally, when push finishes it needs to refresh the ECR repository in the explorer, but it worked and was easy to use :D

**Comment by hunterwerlla** [comment]
If you cancel the image push, this happens:
Screen Shot 2021-04-27 at 1 08 27 PM
(for some reason that doesn't output anything to idea.log as well)

File: jetbrains-core/it-211+/software/aws/toolkits/jetbrains/services/ecr/EcrPushIntegrationTest.kt [hunk 9]
```diff
@@ -51,21 +59,56 @@ class EcrPushIntegrationTest {
...
+
+        val ecrLogin = ecrClient.authorizationToken.authorizationData().first().getDockerLogin()
+        val config = dockerRunConfigurationFromPath(projectRule.project, remoteTag, dockerfile.absolutePath)
+        val pushRequest = DockerfileEcrPushRequest(
+            config.configuration as DockerRunConfiguration,
+            remoteRepo,
+            remoteTag
+        )
+        runBlocking {
+            pushImage(projectRule.project, ecrLogin, pushRequest)
```
Comment (hunterwerlla) [inline jetbrains-core/it-211+/software/aws/toolkits/jetbrains/services/ecr/EcrPushIntegrationTest.kt:111]: nit: I think we can remove the comment in the assert below, we are already testing it above with a different path

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/EcrUtils.kt [hunk 10]
```diff
@@ -32,48 +54,146 @@ data class EcrLogin(
...
+    val serverConnection: ServerConnection<DeploymentConfiguration>,
+    val runtimeInstance: DockerServerRuntimeInstance
+)
+
+private fun getTemporaryDockerConnection() = ServerConnectionManager.getInstance().createTemporaryConnection(
+    RemoteServerImpl("DockerConnection", DockerCloudType.getInstance(), DockerCloudConfiguration.createDefault())
 )
 
-suspend fun getDockerServerRuntimeInstance(): DockerServerRuntimeInstance {
+suspend fun getDockerServerRuntimeInstance(existingServer: RemoteServer<DockerCloudConfiguration>? = null): DockerConnection {
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/EcrUtils.kt:80]: nit: why not just `server`?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/EcrUtils.kt [hunk 11]
```diff
@@ -32,48 +54,146 @@ data class EcrLogin(
...
-    val (username, password) = ecrLogin
-    val model = DockerRepositoryModel().also {
-        val repoUri = pushRequest.remoteRepo.repositoryUri
-        it.registry = DockerRegistry().also { registry ->
-            registry.address = repoUri
-            registry.username = username
-            registry.password = password
+suspend fun pushImage(project: Project, ecrLogin: EcrLogin, pushRequest: EcrPushRequest) {
+    when (pushRequest) {
+        is DockerfileEcrPushRequest -> buildAndPushDockerfile(project, ecrLogin, pushRequest)
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/EcrUtils.kt:106]: Should we add a log before this like we do in the ImageEcrPushRequest path?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/EcrUtils.kt [hunk 12]
```diff
@@ -32,48 +54,146 @@ data class EcrLogin(
...
+        runConfiguration.deploymentConfiguration,
+        project,
+        null,
+        execEnviron
+    )
+
+    // check we don't have something running
+    val expectedDeploymentName = dockerRuntime.getDeploymentName(runConfiguration.deploymentSource, runConfiguration.deploymentConfiguration)
+    if (serverConnection.deployments.firstOrNull { it.name == expectedDeploymentName && it.status == DeploymentStatus.DEPLOYING } != null) {
+        notifyError("title", "deployment currently in progress")
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/EcrUtils.kt:150]: nit: needs localization (including the notifyError's below)

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/EcrUtils.kt [hunk 13]
```diff
@@ -32,48 +54,146 @@ data class EcrLogin(
...
-    val dockerApplicationRuntime = getDockerApplicationRuntimeInstance(dockerServerRuntime, pushRequest.localImageId)
-    dockerApplicationRuntime.pushImage(project, model)
+    // seems gross but this is cleaner than manually attempting to expose the build logs to the user
+    val deployment = deploymentPromise.await()
+    // TODO: why doesn't logging to this log handler do anything?
+    val logHandler = deployment.getOrCreateLogManager(project).mainLoggingHandler
+
+    // unfortunately no callbacks available to grab failure
+    while (deployment.status == DeploymentStatus.DEPLOYING) {
+        delay(500)
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/EcrUtils.kt:171]: nit: `100` instead of `500`?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/EcrUtils.kt [hunk 14]
```diff
@@ -102,3 +222,35 @@ fun SdkRepository.toToolkitEcrRepository(): Repository? {
 
     return Repository(name, arn, uri)
 }
+
+object EcrUtils {
+    val LOG = getLogger<EcrUtils>()
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/EcrUtils.kt:227]: I'm not sure if it would be better to leave the functions as top level functions or put them in the `object`, I'm leaning towards the second because they all use `LOG`

**Comment by rli** [comment]
> If you cancel the image push, this happens:
> (for some reason that doesn't output anything to idea.log as well)

youtrack: https://youtrack.jetbrains.com/issue/IDEA-268237

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushToRepositoryAction.kt [hunk 15]
```diff
@@ -0,0 +1,360 @@
...
+            }
+
+        model.replaceAll(configs)
+        model.selectedItem = configs.firstOrNull()
+    }
+
+    private fun selectedRepo() = remoteRepos.selected() ?: throw IllegalStateException("repository uri was null")
+
+    suspend fun getPushRequest(): EcrPushRequest {
+        val tag = remoteTag.nullize() ?: defaultTag
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushToRepositoryAction.kt:337]: Should we also take into account blank? I guess a user did intend it if it is blank though

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushToRepositoryAction.kt [hunk 16]
```diff
@@ -0,0 +1,360 @@
...
+                    listCellRenderer { value, _, _ ->
+                        icon = value.icon
+                        text = value.name
+                    }
+                )
+                    .growPolicy(GrowPolicy.MEDIUM_TEXT)
+                    .withErrorOnApplyIf(message("ecr.dockerfile.configuration.invalid")) { it.selected() == null }
+                    .withErrorOnApplyIf(message("ecr.dockerfile.configuration.invalid_server")) { it.selected()?.serverName == null }
+
+                // TODO: how do we render both the Docker icon and action items correctly?
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushToRepositoryAction.kt:253]: This todo was fulfilled right? at least it looked correct when I tried it

**Review by hunterwerlla — APPROVED** [review]

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushToRepositoryAction.kt (hunk 16 above)
Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushToRepositoryAction.kt:253]: image

my goal was to have the Docker icon show up, but current state feels good enough for now

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushToRepositoryAction.kt (hunk 15 above)
Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushToRepositoryAction.kt:342]: added validation

**Review by hunterwerlla — APPROVED** [review]

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushToRepositoryAction.kt [hunk 17]
```diff
@@ -0,0 +1,365 @@
...
+    CoroutineScope by ApplicationThreadPoolScope("PushRepositoryAction") {
+    private val dockerServerRuntime: Deferred<DockerServerRuntimeInstance> =
+        async(start = CoroutineStart.LAZY) { EcrUtils.getDockerServerRuntimeInstance().runtimeInstance }
+
+    override fun actionPerformed(e: AnActionEvent) {
+        val project = e.getRequiredData(LangDataKeys.PROJECT)
+        val client: EcrClient = project.awsClient()
+
+        val selectedRepository = e.getData(ExplorerDataKeys.SELECTED_NODES)
+            ?.mapNotNull { it as? EcrRepositoryNode }
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushToRepositoryAction.kt:83]: If you select 2 different types of nodes, shold we still show ECR? I think not, so maybe the take If should go first?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushToRepositoryAction.kt [hunk 18]
```diff
@@ -0,0 +1,365 @@
...
+
+    private fun dockerfileConfigurationSelectorPanel() = panel {
+        row(message("ecr.dockerfile.configuration.label")) {
+            cell {
+                val model = CollectionComboBoxModel<DockerRunConfiguration>()
+                rebuildRunConfigurationComboBoxModel(model)
+                val box = comboBox(
+                    model,
+                    { runConfiguration },
+                    { ::runConfiguration.set(it) },
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushToRepositoryAction.kt:248]: Why is this not using `::runConfiguration` passed to `comboBox(
    model: ComboBoxModel<T>,
    prop: KMutableProperty0<T>,
    renderer: ListCellRenderer<T?>? = null
  ): CellBuilder<ComboBox<T>>`

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushToRepositoryAction.kt [hunk 19]
```diff
@@ -0,0 +1,365 @@
...
+    private val dockerServerRuntime: Deferred<DockerServerRuntimeInstance>
+) : DialogWrapper(project, null, false, IdeModalityType.PROJECT),
+    CoroutineScope by ApplicationThreadPoolScope("PushRepositoryDialog") {
+    private val defaultTag = "latest"
+    private var type: BuildType = BuildType.LocalImage
+    private var remoteTag: String = ""
+    private val localImageRepoTags = CollectionComboBoxModel<LocalImage>()
+
+    private var localImage: LocalImage? = null
+    private var runConfiguration: DockerRunConfiguration? = null
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushToRepositoryAction.kt:155]: Can we be consistent on specifying type vs not?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushToRepositoryAction.kt (hunk 18 above)
Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushToRepositoryAction.kt:248]: Oh, I see the comment earlier in file

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushToRepositoryAction.kt (hunk 17 above)
Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushToRepositoryAction.kt:83]: sure

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushToRepositoryAction.kt (hunk 18 above)
Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushToRepositoryAction.kt:248]: yeah it appears to only happen with the nullable fields, and the whole system around those is a bit wonky so it's actually cleaner like this

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushToRepositoryAction.kt (hunk 19 above)
Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushToRepositoryAction.kt:155]: fixed

**Review by abrooksv — APPROVED** [review]
